File: src-tauri/src/api/anilist.rs
use reqwest::Client;
use serde_json::{json, Value};

const ENDPOINT: &str = "https://graphql.anilist.co";

// AniList caps Page.perPage at 50; we use 20 to match the other providers.
const PAGE_SIZE: u32 = 20;

// Shared media fields used by both list and search queries.
const MEDIA_LIST_FIELDS: &str = r#"
    id
    type
    format
    status
    episodes
    chapters
    volumes
    averageScore
    popularity
    genres
    bannerImage
    coverImage { extraLarge large medium }
    title { romaji english native userPreferred }
    startDate { year month day }
    description(asHtml: false)
"#;

const MEDIA_DETAIL_FIELDS: &str = r#"
    id
    type
    format
    status
    episodes
    chapters
    volumes
    duration
    averageScore
    meanScore
    popularity
    favourites
    genres
    bannerImage
    coverImage { extraLarge large medium }
    title { romaji english native userPreferred }
    startDate { year month day }
    endDate { year month day }
    season
    seasonYear
    description(asHtml: false)
    siteUrl
    studios(isMain: true) { nodes { id name } }
    staff(perPage: 6) { edges { role node { id name { full } } } }
    characters(perPage: 20, sort: ROLE) {
      edges {
        role
        node { id name { full } image { large medium } }
      }
    }
    trailer { id site thumbnail }
"#;

fn http() -> Client {
    Client::builder()
        .user_agent("TauriFlix/1.0")
        .build()
        .unwrap_or_else(|_| Client::new())
}

// AniList returns errors in `{ errors: [{ message }] }`. Surface them as Err.
fn check_graphql(v: &Value) -> Result<(), String> {
    if let Some(arr) = v.get("errors").and_then(|e| e.as_array()) {
        if !arr.is_empty() {
            let msg = arr
                .iter()
                .filter_map(|e| e.get("message").and_then(|m| m.as_str()))
                .collect::<Vec<_>>()
                .join("; ");
            return Err(format!("AniList: {}", msg));
        }
    }
    Ok(())
}

async fn graphql(query: &str, variables: Value) -> Result<Value, String> {
    let body = json!({ "query": query, "variables": variables });
    let res = http()
        .post(ENDPOINT)
        .header("Accept", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| {
            eprintln!("[anilist] ERROR: {e}");
            e.to_string()
        })?
        .json::<Value>()
        .await
        .map_err(|e| e.to_string())?;
    check_graphql(&res)?;
    Ok(res)
}

// Build the list/search query. When no search term is provided we sort by
// popularity so the discover view is meaningful (mirrors the other providers).
fn list_query(media_type: &str) -> String {
    format!(
        r#"
        query ($page: Int, $perPage: Int, $search: String, $genre: String, $sort: [MediaSort]) {{
          Page(page: $page, perPage: $perPage) {{
            pageInfo {{ total currentPage lastPage hasNextPage perPage }}
            media(type: {media_type}, search: $search, genre: $genre, sort: $sort, isAdult: false) {{
              {fields}
            }}
          }}
        }}
        "#,
        media_type = media_type,
        fields = MEDIA_LIST_FIELDS
    )
}

fn detail_query() -> String {
    format!(
        r#"
        query ($id: Int) {{
          Media(id: $id) {{
            {fields}
          }}
        }}
        "#,
        fields = MEDIA_DETAIL_FIELDS
    )
}

async fn run_list(
    media_type: &str,
    query: &str,
    page: u32,
    genre: Option<String>,
) -> Result<Value, String> {
    let has_query = !query.trim().is_empty();
    // Search relevance handles ordering when the user typed something;
    // otherwise we explicitly request popularity-desc.
    let sort: Vec<&str> = if has_query {
        vec!["SEARCH_MATCH"]
    } else {
        vec!["POPULARITY_DESC"]
    };

    let mut vars = json!({
        "page": page,
        "perPage": PAGE_SIZE,
        "sort": sort,
    });
    if has_query {
        vars["search"] = json!(query);
    }
    if let Some(g) = genre.as_ref().filter(|s| !s.is_empty()) {
        vars["genre"] = json!(g);
    }

    graphql(&list_query(media_type), vars).await
}

// ── ANIME ─────────────────────────────────────────────────
#[tauri::command]
pub async fn anilist_search_anime(
    query: &str,
    page: u32,
    genre: Option<String>,
) -> Result<Value, String> {
    eprintln!(
        "[anilist] search_anime  query={:?} page={} genre={:?}",
        query, page, genre
    );
    run_list("ANIME", query, page, genre).await
}

#[tauri::command]
pub async fn anilist_anime_details(id: u32) -> Result<Value, String> {
    eprintln!("[anilist] anime_details  id={}", id);
    let res = graphql(&detail_query(), json!({ "id": id })).await?;
    res.get("data")
        .and_then(|d| d.get("Media"))
        .cloned()
        .ok_or_else(|| "Anime não encontrado.".to_string())
}

// ── MANGA ─────────────────────────────────────────────────
#[tauri::command]
pub async fn anilist_search_manga(
    query: &str,
    page: u32,
    genre: Option<String>,
) -> Result<Value, String> {
    eprintln!(
        "[anilist] search_manga  query={:?} page={} genre={:?}",
        query, page, genre
    );
    run_list("MANGA", query, page, genre).await
}

#[tauri::command]
pub async fn anilist_manga_details(id: u32) -> Result<Value, String> {
    eprintln!("[anilist] manga_details  id={}", id);
    let res = graphql(&detail_query(), json!({ "id": id })).await?;
    res.get("data")
        .and_then(|d| d.get("Media"))
        .cloned()
        .ok_or_else(|| "Mangá não encontrado.".to_string())
}

// ── GENRES ────────────────────────────────────────────────
// AniList exposes a single `GenreCollection` shared by anime and manga.
#[tauri::command]
pub async fn anilist_genres() -> Result<Value, String> {
    eprintln!("[anilist] genres");
    let res = graphql("query { GenreCollection }", json!({})).await?;
    Ok(res
        .get("data")
        .and_then(|d| d.get("GenreCollection"))
        .cloned()
        .unwrap_or(Value::Array(vec![])))
}

File: src/lib/api/anilist.ts
// ================================================================
// TauriFlix — AniList Service (Anime + Manga)
//
// AniList is GraphQL-based; the Rust backend issues the actual HTTP POST
// and returns the unwrapped GraphQL `data.Page` / `data.Media` payloads.
// AniList shares the same `GenreCollection` for anime and manga, so both
// service methods point at the same underlying command.
// ================================================================
import { invoke } from '@tauri-apps/api/core';
import type { MediaItem, MediaDetail, PaginatedResult, GenreOption, MediaType } from '$lib/types/media';

interface RawAnilistPage {
  data: {
    Page: {
      pageInfo: { total: number; currentPage: number; lastPage: number; hasNextPage: boolean; perPage: number };
      media: any[];
    };
  };
}

// Pick the most appropriate localised title.
function pickTitle(title: any): string {
  if (!title) return 'Sem título';
  return title.userPreferred || title.english || title.romaji || title.native || 'Sem título';
}

function pickPoster(cover: any): string | null {
  if (!cover) return null;
  return cover.extraLarge || cover.large || cover.medium || null;
}

function fuzzyDate(d: any): string {
  if (!d || !d.year) return '';
  const m = String(d.month ?? 1).padStart(2, '0');
  const day = String(d.day ?? 1).padStart(2, '0');
  return `${d.year}-${m}-${day}`;
}

// AniList scores are 0-100; convert to the 0-10 scale used elsewhere.
function score10(s: number | null | undefined): number {
  return typeof s === 'number' && s > 0 ? +(s / 10).toFixed(1) : 0;
}

function stripHtml(s: string | null | undefined): string {
  if (!s) return '';
  return s.replace(/<[^>]+>/g, '').replace(/\s+/g, ' ').trim();
}

function mapMedia(raw: any, mt: MediaType): MediaItem {
  return {
    id:            raw.id,
    title:         pickTitle(raw.title),
    overview:      stripHtml(raw.description),
    poster_path:   pickPoster(raw.coverImage),
    backdrop_path: raw.bannerImage ?? null,
    vote_average:  score10(raw.averageScore),
    vote_count:    raw.popularity ?? 0,
    release_date:  fuzzyDate(raw.startDate),
    media_type:    mt,
    episodes:      mt === 'anime' ? (raw.episodes ?? null) : undefined,
    chapters:      mt === 'manga' ? (raw.chapters ?? null) : undefined,
  };
}

function mapPage(raw: RawAnilistPage, mt: MediaType): PaginatedResult<MediaItem> {
  const p = raw?.data?.Page;
  const list = p?.media ?? [];
  return {
    results:       list.map(m => mapMedia(m, mt)),
    page:          p?.pageInfo?.currentPage ?? 1,
    total_pages:   p?.pageInfo?.lastPage ?? 1,
    total_results: p?.pageInfo?.total ?? list.length,
  };
}

function mapDetail(raw: any, mt: MediaType): MediaDetail {
  const cast = (raw.characters?.edges ?? []).slice(0, 20).map((e: any) => ({
    id:           e.node?.id ?? 0,
    name:         e.node?.name?.full ?? '',
    character:    e.role ? String(e.role).replace('_', ' ').toLowerCase() : '',
    profile_path: e.node?.image?.large ?? e.node?.image?.medium ?? null,
  }));

  const studios = Array.isArray(raw.studios?.nodes)
    ? raw.studios.nodes.map((s: any) => s.name).filter(Boolean)
    : [];
  const author = (raw.staff?.edges ?? [])
    .filter((e: any) => /story|art|original/i.test(e.role ?? ''))
    .map((e: any) => e.node?.name?.full)
    .filter(Boolean)
    .join(', ');

  const videos = raw.trailer?.id && raw.trailer?.site === 'youtube'
    ? [{ key: raw.trailer.id, site: 'YouTube', type: 'Trailer', name: 'Trailer' }]
    : [];

  const genres = Array.isArray(raw.genres)
    ? raw.genres.map((g: string) => ({ id: g, name: g }))
    : [];

  return {
    id:            raw.id,
    media_type:    mt,
    title:         pickTitle(raw.title),
    tagline:       '',
    overview:      stripHtml(raw.description),
    poster_path:   pickPoster(raw.coverImage),
    backdrop_path: raw.bannerImage ?? null,
    vote_average:  score10(raw.averageScore ?? raw.meanScore),
    vote_count:    raw.popularity ?? 0,
    release_date:  fuzzyDate(raw.startDate),
    runtime:       raw.duration ?? null,
    genres,
    cast,
    videos,
    episodes:      mt === 'anime' ? (raw.episodes ?? null) : undefined,
    chapters:      mt === 'manga' ? (raw.chapters ?? null) : undefined,
    volumes:       mt === 'manga' ? (raw.volumes ?? null) : undefined,
    status:        raw.status ?? '',
    studios,
    author:        mt === 'manga' ? author : undefined,
  };
}

// Genres are returned as a flat string array shared across anime and manga.
function mapGenres(raw: string[]): GenreOption[] {
  return (raw ?? []).map(name => ({ id: name, name }));
}

export const AnilistAPI = {
  searchAnime: (query: string, page = 1, genre?: string) =>
    invoke<RawAnilistPage>('anilist_search_anime', { query, page, genre: genre ?? null })
      .then(r => mapPage(r, 'anime')),

  searchManga: (query: string, page = 1, genre?: string) =>
    invoke<RawAnilistPage>('anilist_search_manga', { query, page, genre: genre ?? null })
      .then(r => mapPage(r, 'manga')),

  animeDetails: (id: number): Promise<MediaDetail> =>
    invoke<any>('anilist_anime_details', { id }).then(r => mapDetail(r, 'anime')),

  mangaDetails: (id: number): Promise<MediaDetail> =>
    invoke<any>('anilist_manga_details', { id }).then(r => mapDetail(r, 'manga')),

  animeGenres: (): Promise<GenreOption[]> =>
    invoke<string[]>('anilist_genres').then(mapGenres),

  // Same underlying GenreCollection for both — exposed twice to keep the
  // home page's per-category genre cache trivial.
  mangaGenres: (): Promise<GenreOption[]> =>
    invoke<string[]>('anilist_genres').then(mapGenres),
};
